lib/ASTGen/Sources/ASTGen/Types.swift

```diff
@@ -207,3 +207,37 @@ extension ASTGenVisitor {
     }
   }
 }
+
+@_cdecl("swift_ASTGen_buildTypeRepr")
+@usableFromInline
+func buildTypeRepr(
+  sourceFilePtr: UnsafeRawPointer,
+  typeLocPtr: UnsafePointer<UInt8>,
+  dc: UnsafeMutableRawPointer,
+  ctx: UnsafeMutableRawPointer,
+  endTypeLocPtr: UnsafeMutablePointer<UnsafePointer<UInt8>?>
+) -> UnsafeMutableRawPointer? {
+  let sourceFile = sourceFilePtr.bindMemory(
+    to: ExportedSourceFile.self, capacity: 1
+  )
+
+  // Find the type syntax node.
+  guard let typeSyntax = findSyntaxNodeInSourceFile(
+    sourceFilePtr: sourceFilePtr,
+    sourceLocationPtr: typeLocPtr,
+    type: TypeSyntax.self,
+    wantOutermost: true
+  ) else {
+    // FIXME: Produce an error
+    return nil
+  }
+
+  // Fill in the end location.
+  endTypeLocPtr.pointee = sourceFile.pointee.buffer.baseAddress!.advanced(by: typeSyntax.endPosition.utf8Offset)
+
+  // Convert the type syntax node.
+  let typeReprNode = ASTGenVisitor(ctx: ctx, base: sourceFile.pointee.buffer.baseAddress!, declContext: dc)
+    .visit(typeSyntax)
```

Comment:
Out of curiosity: How complete is ASTGen for types?

Reply:
A lot of it is there! There is a huge gap where we drop all of the attributes, which breaks a ton of test cases, and I have a few bug fixes locally to pick up specific things that are missing (`inout` doesn't work, there's a bug with generic argument lists that causes a crash, etc.). I'll push a group of those in a bit.

Reply:
FWIW, I've picked off many of these, but there are still more to do. The biggest remaining grind here is to deal with all of the type attributes that have special rules (`@convention`, `@differentiable`, etc.) and wire up custom attributes.